agent/eventhandler/task_handler.go

```diff
@@ -15,63 +15,80 @@ package eventhandler
 
 import (
 	"container/list"
+	"sync"
 	"time"
 
 	"github.com/aws/amazon-ecs-agent/agent/api"
 	"github.com/aws/amazon-ecs-agent/agent/utils"
 )
 
-var handler *taskHandler
+// Maximum number of tasks that may be handled at once by the taskHandler
+const concurrentEventCalls = 3
 
-func init() {
-	handler = newTaskHandler()
+type eventList struct {
+	sending    bool // whether the list is already being handled
+	sync.Mutex      // Locks both the list and sending bool
+	*list.List      // list of *sendableEvents
+}
+
+type taskHandler struct {
+	submitSemaphore utils.Semaphore       // Semaphore on the number of tasks that may be handled at once
+	taskMap         map[string]*eventList // arn:*eventList map so events may be serialized per task
+	sync.RWMutex                          // Lock for the taskMap
+}
+
+func NewTaskHandler() *taskHandler {
```

Comment:
A global method returning a local data type doesn't seem like the right thing to do in terms of api abstraction and design. Please make `taskHandler` a global type as it has methods that are global as well.